independent-projects/arc/runtime/src/main/java/io/quarkus/arc/AbstractInvocationContext.java

```diff
@@ -3,41 +3,35 @@
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Method;
+import java.util.Arrays;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
-import java.util.function.Function;
 import javax.interceptor.InvocationContext;
 
-public abstract class AbstractInvocationContext implements InvocationContext {
+abstract class AbstractInvocationContext implements InvocationContext {
 
-    protected Map<String, Object> contextData;
+    protected volatile Map<String, Object> contextData;
```

Comment:
Could you add some comments on the need for `volatile` ? 

maybe it's just me being unfamiliar with CDI, but it would highly help me understand the possibilities here with some more code comments :)

Reply:
Well, the `contextData` is initialized lazily, when needed (as it's usually not needed), see the `newContextData()` method, and since it could be accessed from a different thread (see my email about MP FT requirements/problems) we need to make sure it's safely propagated. I'll try to update the javadoc to make it more clear.

Reply:
> we need to make sure it's safely propagated.

The current implementation makes the around invoke (the only candidate to be passed around multiple threads) create a new instance in each step, hence the different thread will get a new instance of IC, which just gets the map as a parameter. I am probably missing some obvious bit as for why we need volatile?

Reply:
You're probably right. I was thinking about accessing the context data from the "previous" IC instance which woudln't work anyway (neither works in Weld imho):
```
--> INT1 - proceed() 
  --> INT2 - proceed() and set some context data
    --> invoke target method
  --> INT2 return from the `@AroundInvoke` method
--> INT1 - access context data after proceed()
```

Reply:
Hmm, I'd guess that `INT1` will only see whatever context data was added before it was invoked or what `INT1` itself added, right?
Then again, looking at the code, we just pass around the same map, so _theoretically_ you could see that? Might be worth trying out...

Reply:
You don't need volatile to continue execution in a different thread.  In order for an object X to be "moved" from thread 1 to thread 2, the object X must be safely published.  Publishing X implicitly publishes everything that was written before doing so, including indexes etc.

You'd only need `volatile` or atomic constructs if thread 1 continued to access object X *after* it was published to thread 2.  A clean handoff means you don't need any other publication or synchronization primitives.

Reply:
> Hmm, I'd guess that INT1 will only see whatever context data was added before it was invoked or what INT1 itself added, right?

@manovotn Yes. And I think we should fix it.

> Then again, looking at the code, we just pass around the same map, so theoretically you could see that?

I don't think so. `IC1.contextData` is null.

> You'd only need volatile or atomic constructs if thread 1 continued to access object X after it was published to thread 2...

@dmlloyd thanks for explanation David. I think that our use case is a bit more complicated, ie. I thought we will need to access object X after it was published to thread 2. But I need to think about it.